From 39d69da5ea02dcd25b98224d40dc00d3f9ee8323 Mon Sep 17 00:00:00 2001
From: Chen Minqiang <ptpt52@gmail.com>
Date: Fri, 11 Oct 2024 14:07:00 +0800
Subject: [PATCH] net: Allocate ifindex from range 1-63 using bitmap

---
 net/core/dev.c | 37 +++++++++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/net/core/dev.c b/net/core/dev.c
index 04a9964c75fa..4e35875b654d 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -10321,6 +10321,16 @@ static void netdev_do_free_pcpu_stats(struct net_device *dev)
 	}
 }
 
+static unsigned int net_device_ifindex_bits[2] = {0, 0};
+static void release_ifindex(int ifindex)
+{
+	if (ifindex < 32) {
+		net_device_ifindex_bits[0] &= ~(1 << ifindex);
+	} else if (ifindex < 64) {
+		net_device_ifindex_bits[1] &= ~(1 << (ifindex - 32));
+	}
+}
+
 /**
  * register_netdevice() - register a network device
  * @dev: device to register
@@ -10389,11 +10399,36 @@ int register_netdevice(struct net_device *dev)
 	if (ret)
 		goto err_uninit;
 
+	do {
+		if (dev->ifindex > 0)
+			break;
+		for (ret = 1; ret < 32; ret++) {
+			if (!(net_device_ifindex_bits[0] & (1 << ret))) {
+				dev->ifindex = ret;
+				break;
+			}
+		}
+		if (dev->ifindex > 0)
+			break;
+		for (ret = 0; ret < 32; ret++) {
+			if (!(net_device_ifindex_bits[1] & (1 << ret))) {
+				dev->ifindex = ret + 32;
+				break;
+			}
+		}
+	} while (0);
+
 	ret = dev_index_reserve(net, dev->ifindex);
 	if (ret < 0)
 		goto err_free_pcpu;
 	dev->ifindex = ret;
 
+	if (dev->ifindex < 32) {
+		net_device_ifindex_bits[0] |= 1 << dev->ifindex;
+	} else if (dev->ifindex < 64) {
+		net_device_ifindex_bits[1] |= 1 << (dev->ifindex - 32);
+	}
+
 	/* Transfer changeable features to wanted_features and enable
 	 * software offloads (GSO and GRO).
 	 */
@@ -10496,6 +10531,7 @@ int register_netdevice(struct net_device *dev)
 err_uninit_notify:
 	call_netdevice_notifiers(NETDEV_PRE_UNINIT, dev);
 err_ifindex_release:
+	release_ifindex(dev->ifindex);
 	dev_index_release(net, dev->ifindex);
 err_free_pcpu:
 	netdev_do_free_pcpu_stats(dev);
@@ -11283,6 +11319,7 @@ void unregister_netdevice_queue(struct net_device *dev, struct list_head *head)
 	if (head) {
 		list_move_tail(&dev->unreg_list, head);
 	} else {
+		release_ifindex(dev->ifindex);
 		LIST_HEAD(single);
 
 		list_add(&dev->unreg_list, &single);
-- 
2.17.1

