--- a/drivers/mtd/spi-nor/core.c
+++ b/drivers/mtd/spi-nor/core.c
@@ -3062,7 +3062,7 @@ static const struct flash_info *spi_nor_
 	return NULL;
 }
 
-static int spi_nor_set_addr_width(struct spi_nor *nor)
+static int spi_nor_set_addr_width(struct device_node *np, struct spi_nor *nor)
 {
 	if (nor->addr_width) {
 		/* already configured from SFDP */
@@ -3077,6 +3077,13 @@ static int spi_nor_set_addr_width(struct
 		nor->addr_width = 4;
 	}
 
+	if (of_property_read_bool(np, "spi-3byte-addressing")) {
+		if (nor->mtd.size > 0x1000000) {
+			nor->addr_width = 3;
+			nor->mtd.size = 0x1000000;
+		}
+	}
+
 	if (nor->addr_width > SPI_NOR_MAX_ADDR_WIDTH) {
 		dev_dbg(nor->dev, "address width is too large: %u\n",
 			nor->addr_width);
@@ -3263,7 +3270,7 @@ int spi_nor_scan(struct spi_nor *nor, co
 	if (info->flags & SPI_NOR_4B_OPCODES)
 		nor->flags |= SNOR_F_4B_OPCODES;
 
-	ret = spi_nor_set_addr_width(nor);
+	ret = spi_nor_set_addr_width(np, nor);
 	if (ret)
 		return ret;
 
